pandas/core/dtypes/concat.py

```diff
@@ -132,6 +183,73 @@ def is_nonempty(x) -> bool:
     return np.concatenate(to_concat, axis=axis)
 
 
+def concat_arrays(to_concat: List[Any]) -> ArrayLike:
+    """
+    Alternative for concat_compat but specialized for use in the ArrayManager.
+
+    Differences: only deals with 1D arrays (no axis keyword) and does not skip
+    empty arrays to determine the dtype.
+    In addition ensures that all NullArrayProxies get replaced with actual
+    arrays.
+
+    Parameters
+    ----------
+    to_concat : list of arrays
+
+    Returns
+    -------
+    np.ndarray or ExtensionArray
+    """
+    # ignore the all-NA proxies to determine the resulting dtype
+    to_concat_no_proxy = [x for x in to_concat if not isinstance(x, NullArrayProxy)]
+
+    kinds = {obj.dtype.kind for obj in to_concat_no_proxy}
+    single_dtype = len({x.dtype for x in to_concat_no_proxy}) == 1
+    any_ea = any(is_extension_array_dtype(x.dtype) for x in to_concat_no_proxy)
+
+    if any_ea:
+        if not single_dtype:
+            target_dtype = find_common_type([x.dtype for x in to_concat_no_proxy])
+            to_concat = [_cast_to_common_type(arr, target_dtype) for arr in to_concat]
+        else:
+            target_dtype = to_concat_no_proxy[0].dtype
+            to_concat = [
+                arr.to_array(target_dtype) if isinstance(arr, NullArrayProxy) else arr
+                for arr in to_concat
+            ]
+
+        if isinstance(to_concat[0], ExtensionArray):
+            cls = type(to_concat[0])
+            return cls._concat_same_type(to_concat)
+        else:
+            return np.concatenate(to_concat)
+
+    elif any(kind in ["m", "M"] for kind in kinds):
+        return _concat_datetime(to_concat)
+
+    if not single_dtype:
+        target_dtype = np.find_common_type(
+            [arr.dtype for arr in to_concat_no_proxy], []
+        )
+    else:
+        target_dtype = to_concat_no_proxy[0].dtype
+    to_concat = [
+        arr.to_array(target_dtype) if isinstance(arr, NullArrayProxy) else arr
+        for arr in to_concat
+    ]
+
+    result = np.concatenate(to_concat)
+
+    # TODO(ArrayManager) this is currently inconsistent between Series and DataFrame
+    # so we should decide whether to keep the below special case or remove it
```

Comment:
is this behavior consistent between AM and BM?

Reply:
Ah, found my notes about this: this is not directly related to Block vs ArrayManager, but something I noticed here that is inconsistent between Series vs DataFrame and empty vs non-empty. 

For Series, with non-empty, we actually coerce to numeric, while if we have empty bool + float, we get object dtype:

```
In [7]: pd.concat([pd.Series([True], dtype=bool), pd.Series([1.0], dtype=float)])
Out[7]: 
0    1.0
0    1.0
dtype: float64

In [8]: pd.concat([pd.Series([], dtype=bool), pd.Series([], dtype=float)])
Out[8]: Series([], dtype: object)
```

While DataFrame uses object dtype in both cases.

(will open a separate issue about it)

But so the reason this came up specifically for ArrayManager, is that we now follow the same rules for Series/DataFrame, and thus also get numeric for non-empty for the DataFrame case.

Reply:
Opened https://github.com/pandas-dev/pandas/issues/39817 about this